File: Src/java/engine/src/main/java/org/opencds/cqf/cql/engine/elm/execution/DivideEvaluator.java
package org.opencds.cqf.cql.engine.elm.execution;

import java.math.BigDecimal;
import java.math.RoundingMode;

import org.opencds.cqf.cql.engine.exception.InvalidOperatorArgument;
import org.opencds.cqf.cql.engine.execution.Context;
import org.opencds.cqf.cql.engine.runtime.Interval;
import org.opencds.cqf.cql.engine.runtime.Quantity;
import org.opencds.cqf.cql.engine.runtime.Value;

/*
/(left Decimal, right Decimal) Decimal
/(left Quantity, right Decimal) Quantity
/(left Quantity, right Quantity) Quantity

The divide (/) operator performs numeric division of its arguments.
Note that this operator is Decimal division; for Integer division, use the truncated divide (div) operator.
When invoked with Integer arguments, the arguments will be implicitly converted to Decimal.
TODO: For division operations involving quantities, the resulting quantity will have the appropriate unit. For example:
12 'cm2' / 3 'cm'
In this example, the result will have a unit of 'cm'.
If either argument is null, the result is null.
*/

public class DivideEvaluator extends org.cqframework.cql.elm.execution.Divide {

    private static BigDecimal divideHelper(BigDecimal left, BigDecimal right, Context context) {
        if (EqualEvaluator.equal(right, new BigDecimal("0.0"), context)) {
            return null;
        }

        try {
            return Value.verifyPrecision(left.divide(right), null);
        } catch (ArithmeticException e) {
            return left.divide(right, 8, RoundingMode.FLOOR);
        }
    }

    public static Object divide(Object left, Object right, Context context) {

        if (left == null || right == null) {
            return null;
        }

        if (left instanceof BigDecimal && right instanceof BigDecimal) {
            return divideHelper((BigDecimal) left, (BigDecimal) right, context);
        }

        else if (left instanceof Quantity && right instanceof Quantity) {
            BigDecimal value = divideHelper(((Quantity) left).getValue(), ((Quantity) right).getValue(), context);
            return new Quantity().withValue(Value.verifyPrecision(value, null)).withUnit(((Quantity) left).getUnit());
        }

        else if (left instanceof Quantity && right instanceof BigDecimal) {
            BigDecimal value = divideHelper(((Quantity) left).getValue(), (BigDecimal) right, context);
            return new Quantity().withValue(Value.verifyPrecision(value, null)).withUnit(((Quantity)left).getUnit());
        }

        else if (left instanceof Interval && right instanceof Interval) {
            Interval leftInterval = (Interval)left;
            Interval rightInterval = (Interval)right;

            return new Interval(
                    divide(leftInterval.getStart(), rightInterval.getStart(), context), true,
                    divide(leftInterval.getEnd(), rightInterval.getEnd(), context), true
            );
        }

        throw new InvalidOperatorArgument(
                "Divide(Decimal, Decimal), Divide(Quantity, Decimal), Divide(Quantity, Quantity)",
                String.format("Divide(%s, %s)", left.getClass().getName(), right.getClass().getName())
        );
    }

    @Override
    protected Object internalEvaluate(Context context) {
        Object left = getOperand().get(0).evaluate(context);
        Object right = getOperand().get(1).evaluate(context);

        return divide(left, right, context);
    }
}

File: Src/java/engine/src/commonMain/kotlin/org/opencds/cqf/cql/engine/elm/executing/DivideEvaluator.kt
package org.opencds.cqf.cql.engine.elm.executing

import kotlin.jvm.JvmStatic
import org.cqframework.cql.shared.BigDecimal
import org.cqframework.cql.shared.RoundingMode
import org.opencds.cqf.cql.engine.exception.InvalidOperatorArgument
import org.opencds.cqf.cql.engine.execution.State
import org.opencds.cqf.cql.engine.runtime.Interval
import org.opencds.cqf.cql.engine.runtime.Quantity
import org.opencds.cqf.cql.engine.runtime.Value
import org.opencds.cqf.cql.engine.util.javaClassName

/*
/(left Decimal, right Decimal) Decimal
/(left Quantity, right Decimal) Quantity
/(left Quantity, right Quantity) Quantity

The divide (/) operator performs numeric division of its arguments.
Note that this operator is Decimal division; for Integer division, use the truncated divide (div) operator.
When invoked with Integer arguments, the arguments will be implicitly converted to Decimal.
For division operations involving quantities, the resulting quantity will have the appropriate unit. For example:
12 'cm2' / 3 'cm'
In this example, the result will have a unit of 'cm'.
If either argument is null, the result is null.
*/
@Suppress("CyclomaticComplexMethod")
object DivideEvaluator {
    private fun divideHelper(left: BigDecimal, right: BigDecimal?, state: State?): BigDecimal? {
        if (EqualEvaluator.equal(right, BigDecimal("0.0"), state) == true) {
            return null
        }

        return try {
            Value.verifyPrecision(left.divide(right!!), null)
        } catch (e: ArithmeticException) {
            left.divide(right!!, 8, RoundingMode.FLOOR)
        }
    }

    @JvmStatic
    fun divide(left: Any?, right: Any?, state: State?): Any? {
        if (left == null || right == null) {
            return null
        }

        if (left is BigDecimal && right is BigDecimal) {
            return divideHelper(left, right, state)
        } else if (left is Quantity && right is Quantity) {
            val leftValue = left.value!!
            val leftUnit = left.unit!!
            val rightValue = right.value!!
            val rightUnit = right.unit!!
            if (EqualEvaluator.equal(rightValue, BigDecimal("0.0"), state) != false) {
                return null
            }
            val resultValue: BigDecimal?
            val resultUnit: String
            if (rightUnit == "1") {
                resultValue = divideHelper(leftValue, rightValue, state)
                resultUnit = leftUnit
            } else {
                val ucumService = state?.environment?.libraryManager?.ucumService!!
                try {
                    val result =
                        ucumService.divideBy(Pair(leftValue, leftUnit), Pair(rightValue, rightUnit))
                    val unverifiedResultValue = result.first
                    resultValue = Value.verifyPrecision(unverifiedResultValue, null)
                    val rawResultUnit = result.second
                    resultUnit = rawResultUnit.ifEmpty { "1" }
                } catch (e: Exception) {
                    @Suppress("TooGenericExceptionThrown") throw RuntimeException(e)
                }
            }
            return Quantity().withValue(resultValue).withUnit(resultUnit)
        } else if (left is Quantity && right is BigDecimal) {
            val value = divideHelper(left.value!!, right, state)
            return if (value != null) Quantity().withValue(value).withUnit(left.unit) else null
        } else if (left is Interval && right is Interval) {
            return Interval(
                divide(left.start, right.start, state),
                true,
                divide(left.end, right.end, state),
                true,
                state,
            )
        }

        throw InvalidOperatorArgument(
            "Divide(Decimal, Decimal), Divide(Quantity, Decimal), Divide(Quantity, Quantity)",
            "Divide(${left.javaClassName}, ${right.javaClassName})",
        )
    }
}
